q2_assembly/quast/tests/test_quast.py

```diff
@@ -705,6 +732,66 @@ def test_evaluate_contigs_action_paired_end_no_icarus(self, p1, p2, p3, p4):
         }
         p2.assert_called_once_with(ANY, self._tmp, context=exp_context)
 
+    @patch("pandas.read_csv")
+    @patch("q2_assembly.quast._parse_columns")
+    def test_create_tabular_results(self, p1, p2):
+        temp_dir = MockTempDir()
+        report_path = os.path.join(
+            temp_dir.name, "combined_reference", "transposed_report.tsv"
+        )
+        mock_df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
+        p2.return_value = mock_df
+
+        _ = _create_tabular_results(temp_dir.name)
+
+        p2.assert_called_once_with(report_path, sep="\t", header=0)
+        p1.assert_called_once_with(mock_df)
+
+    def test_evaluate_contigs_pipeline(self):
+        # this is used for mocking
+        def _copy_tsv_file(src, dst_dir):
+            if not os.path.exists(dst_dir):
+                os.makedirs(dst_dir)
+
+            destination_path = os.path.join(dst_dir, "quast_results.tsv")
+            shutil.copy2(src, destination_path)
+
+        input_files = self.get_data_path("contigs")
+        _input = ContigSequencesDirFmt(input_files, mode="r")
+        contigs = Artifact.import_data("SampleData[Contigs]", _input)
+
+        enhanced_tabular_results_path = os.path.join(
+            self.get_data_path("quast-results"), "enhanced_tabular_results.tsv"
+        )
+
+        with tempfile.TemporaryDirectory() as tmp:
+            test_enhance_path = os.path.join(tmp, "vis_files", "quast_data")
+            exp_vis = MagicMock()
+            # mocking the _visualize_quast from the ctx.get_action
+            mock_action = MagicMock(side_effect=lambda x, **kwargs: (exp_vis,))
+
+            # mocking the context
+            mock_ctx = MagicMock(
+                get_action=lambda action_type, action_name: mock_action
+            )
+            _copy_tsv_file(enhanced_tabular_results_path, test_enhance_path)
+
+            with patch("tempfile.TemporaryDirectory") as mock_temp_dir:
+                mock_temp_dir.return_value.__enter__.return_value = tmp
+                tab_report, _ = evaluate_contigs(ctx=mock_ctx, contigs=contigs)
+
+                # here I pretend that the tab_report is the dataframe,
+                # so I read it again. The problem is that
+                # ctx.make_artifact is mocked and cannot return the QUASTResults
+                # object.
+                tab_report = pd.read_csv(
+                    enhanced_tabular_results_path, sep="\t", header=0
+                )
+                tab_report_quast = Artifact.import_data("QUASTResults", tab_report)
+
+                tab_report_quast.validate()
+                self.assertIs(tab_report_quast.format, QUASTResultsDirectoryFormat)
```

Comment:
I'm really unsure what to think about this (again)... I feel like this is not really testing anything (except of the fact whether the format is correct but that should anyway be tested somewhere else already). Now that I see it, though, I'm wondering whether we could just run the `evaluate_contigs` as a real action and just mock the `_visualize_quast` method - that way your report would be turned into an artifact by `make_artifact` and you could assert whether it's what you expected. What do you think?

Reply:
This is what I tried to do in the beginning Michal, but somehow it would not work because the `_visualize_quast` is retrieved from the context and the mock would not recognize it as a method being mocked. The code as it stands, does exactly what you explain above, except that unexpectedly it mocks the `make_artifact` as well. If you have any other suggestions on this and we can make it work, it would be ideal. The other way that I was thinking is to mock nothing and test the whole pipeline.

Reply:
Hey @DorielaGrabocka, I played around with that test a little and made it work - please check out the changes in the commit I just pushed :)

Reply:
Thanks Michal!